### PR TITLE
DAOS-16082 test: Repeatedly query device after SysXS device is set to…

### PR DESCRIPTION
**Pull request by shimizukko**
… faulty.

After SysXS device is set to faulty, the device state transition may take a while until its state becomes EVICTED. Wait for 10 sec before querying the device state. If the state isn't EVICTED, query again.

Also set only one SysXS device faulty because there's no point of setting second SysXS device to faulty. (based on the developer feedback).

Skip-unit-tests: true
Skip-fault-injection-test: true
Skip-func-hw-test-medium-md-on-ssd: false
Test-tag: test_dmg_storage_query_device_state

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by phender — CHANGES_REQUESTED** [review]
This test needs to support both PMEM and MD on SSD configurations and the current code already handles this and passes.  The problem is when we run the dmg system query in teardown and we detect `"state": "joined"` because we have not given enough time for the state change to occur.  The simple fix is to add a `dmg system query` loop at the end of the `test_dmg_storage_query_device_state` test to ensure the state transition has occurred before ending the test.  This additional check is only needed if `expect_failed_engine` is set (meaning we set a device w/ sys_xs faulty).

I'd suggest using `wait_for_result()` and defining its `get_method` as a simple method that does something like:
```
def check_engine_down(dmg):
    return not check_system_query_status(dmg.system_query())

**Comment by shimizukko** [comment]
> This test needs to support both PMEM and MD on SSD configurations and the current code already handles this and passes. The problem is when we run the dmg system query in teardown and we detect `"state": "joined"` because we have not given enough time for the state change to occur. The simple fix is to add a `dmg system query` loop at the end of the `test_dmg_storage_query_device_state` test to ensure the state transition has occurred before ending the test. This additional check is only needed if `expect_failed_engine` is set (meaning we set a device w/ sys_xs faulty).
> 
> I'd suggest using `wait_for_result()` and defining its `get_method` as a simple method that does something like:
> 
> ```
> def check_engine_down(dmg):
>     return not check_system_query_status(dmg.system_query())
> ```

Thanks. I updated the test by adding the `dmg system query` check at the end (only when SysXS device was set to faulty.)

File: src/tests/ftest/control/dmg_storage_query.py [hunk 1]
```diff
@@ -247,11 +256,14 @@ def test_dmg_storage_query_device_state(self):
             except CommandFailure:
                 if not expect_failed_engine:
                     self.fail("Error setting the faulty state for {}".format(device['uuid']))
+            # Set only one SysXS device faulty.
+            if expect_failed_engine:
+                break
```
[the remark this answers is not in the document]
Comment (shimizukko) [reply src/tests/ftest/control/dmg_storage_query.py:261]: If we put this `break` after `expect_failed_engine = True` (between line 253 and 254), we will never call set nvme-faulty on the SysXS device.

[the remark this answers is not in the document]
Comment (daltonbohning) [reply src/tests/ftest/control/dmg_storage_query.py:261]: Ah, sorry I misread that

**Review by dinghwah — APPROVED** [review]

**Pull request merged**